man/seq192.1
.\" generated with Ronn-NG/v0.10.1
.\" http://github.com/apjanke/ronn-ng/tree/0.10.1
.TH "SEQ192" "1" "April 2025" "" "User manual"
.SH "NAME"
\fBseq192\fR \- live MIDI sequencer
.SH "SYNOPSIS"
\fBseq192\fR [OPTION\|\.\|\.\|\.]
.SH "DESCRIPTION"
seq192 is a MIDI sequencer for live situations that is meant to be controlled with OSC messages and sync with other applications using jack transport\.
.SH "OPTIONS"
.TP
\fB\-h, \-\-help\fR
Show available options
.TP
\fB\-f, \-\-file\fR \fIfile\fR
Load midi file on startup
.TP
\fB\-c, \-\-config\fR \fIfile\fR
Load config file on startup
.TP
\fB\-p, \-\-osc\-port\fR \fIport\fR
OSC input port (udp port number or unix socket path)
.TP
\fB\-j, \-\-jack\-transport\fR
Sync to jack transport
.TP
\fB\-n, \-\-no\-gui\fR
Enable headless mode
.TP
\fB\-v, \-\-version\fR
Show version and exit
.SH "USER INTERFACE"
Seq192 allows writing and playing MIDI sequences and organizing them in screensets\.
.P
\fBMain window\fR
.P
The main window consist in a toolbar and a sequence grid\.
.IP "\(bu" 4
The \fBtoolbar\fR contains the following controls
.IP
\fIPanic button\fR: disable all sequences
.br
\fIStop button\fR: stop transport
.br
\fIPlay button\fR: start or restart transport
.br
\fIBpm entry\fR: set beats per minute
.br
\fIScreenset name entry\fR: set name of current screenset
.br
\fIScreenset number entry\fR: select current screenset
.br

.IP "\(bu" 4
The \fBsequence grid\fR is empty by default\. Right\-clicking in the grid allows creating a new sequence or editing an existing one\. Sequences can be rearranged in the grid by dragging them\. Left\-clicking on a sequence toggles its playing state\. Middle\-clicking on a sequence opens its edit window\. Shift+click defines the sequence as the synchronization reference for queued sequences\. Ctrl+click toggles the sequence's queued state\. The keyboard's arrow keys can be used to navigate accross sequences\.
.IP "" 0
.P
\fBEdit window\fR
.P
The edit window consist in a menu, a toolbar, a pianoroll and an event editor\.
.IP "\(bu" 4
The \fBtoolbar\fR contains the following controls
.IP
\fISequence name entry\fR: set name of sequence
.br
\fISequence length entries\fR: set number of beats per measure, beat unit and number of measures in sequence
.br
\fISnap button\fR: toggle grid snapping\. Holding \fBalt\fR allows disabling snapping temporarily
.br
\fISnap dropdown\fR: set snapping grid size
.br
\fINote dropdown\fR: set note size
.br
\fIOutput dropdown\fR: set sequence's MIDI port and channel
.br

.IP "\(bu" 4
The \fBpianoroll\fR responds to the following interactions
.IP
\fILeft click (piano area)\fR: send note to sequence's output
.br
\fILeft click\fR: select note(s)
.br
\fILeft click drag\fR: create a lasso selection or move selection
.br
\fIMiddle click (or ctrl + left click) drag\fR: change note(s) duration
.br
\fIRight click + left click\fR: draw new notes
.br

.IP "\(bu" 4
The \fBevent editor\fR allows editing the velocity of notes and adding MIDI events like pitch wheel and control changes to the sequences The button on the left allows selecting which event should be shown on the timeline\. The timeline works like in the piano roll, except events dont have a length\. The bottom part is for editing the events' values: handles can be dragged individually, or one can draw a line with a drag\-and\-drop motion to create a linear automation\. Scrolling with the mousewheel on the event editor increments the values of all selected events\.
.IP "" 0
.P
\fBImportant note\fR
.P
Each window has its own undo/redo history:
.IP "" 4
.nf
\- main window: screnset name and sequences (position and content)
\- edit window: MIDI events, sequence name, number of measures and time signature
.fi
.IP "" 0
.SH "JACK TRANSPORT"
When \fB\-\-jack\-transport\fR is set, seq192 will
.IP "" 4
.nf
\- follow start / stop commands from other clients
\- send start / stop commands to other clients
\- use the transport master's bpm
\- set its position to 0 whenever the transport stops or restarts
\- **not** attempt to reposition within sequences
.fi
.IP "" 0
.SH "CONFIGURATION FILE"
The configration file is located in \fB$XDG_CONFIG_HOME/seq192/config\.json\fR (\fB~/\.config/seq192/config\.json\fR by default), but can be loaded from any location using \fB\-\-config\fR\. It allows customizing the following aspects of seq192:
.IP "" 4
.nf
\- MIDI bus names and portamento settings (see Slide Notes)
\- MIDI channel names per bus
\- Sequence colors as css color strings (per bus or per channel )
\- Note names in the piano roll (per bus or per channel)
\- Control names in the event dropdown (per bus or per channel )
.fi
.IP "" 0
.P
\fBExample\fR
.IP "" 4
.nf
{
    "buses": {
        "0": {
            "name": "Sampler",
            "channels": {
                "0": {
                    "name": "Drums",
                    "color": "orange",
                    "notes": {
                        "64": "Kick",
                        "65": "Snare",
                        "66": "Hihat"
                    },
                    "controls": {
                        "1": "Custom cc name ",
                        "2": "Etc"
                    }
                }
            }
        },
        "1": {
            "name": "Surge XT",
            "portamento_max_time": 4000,
            "portamento_log_scale": true,
            "channels":{
                "0": {"name": "Trap bass"},
                "1": {"name": "Wobble"}
            }
        }
    }
}
.fi
.IP "" 0
.SH "SLIDE NOTES"
Slide notes are an attempt to imitate a well\-known sequencer's feature\. A slide note bends its base note (the highest note that intersects the slide note's beginning) and stops when the base note ends\. The duration of the slide note only defines how long it takes to reach the slide note\.
.P
In the PianoRoll, notes can be toggled from/to slide mode using the "S" key or from the "Edit" menu\.
.P
Slide notes rely on CC5 (Portamento MSB) and CC37 (Portamento LSB) to tell controlled synths their slide duration, this has the following implications
.IP "\(bu" 4
the synth must have portamento enabled
.IP "\(bu" 4
the synth should most likely be monophonic although it may work in some polyphonic scenarios
.IP "\(bu" 4
the synth must bind at least CC5 to its portamento time setting (ideally CC5 and CC37 should be combined to obtain a 14\-bit control)
.IP "\(bu" 4
seq192 must be aware of the synth's maximum portamento time in order to compute correct portamento values
.IP "" 0
.P
Setting the maximum portamento time in seq192 is done using seq192's config file\. Each output bus has a "portamento_max_time" setting (integer, millisecond, defaults to \fB16383\fR) and a "portamento_log_scale" (boolean, defaults to \fBfalse\fR)\.
.P
\fBExamples\fR
.IP "\(bu" 4
\fBFluidsynth\fR: the default portamento setting matches Fluidsynth's portamento implementation\. One should add a CC65 with value 127 (enable Portamento) and a CC67 with value 127 (enable Legato to make it monophonic if needed) at the beginning of the sequence unless these options are already set
.IP "\(bu" 4
\fBSurge XT\fR: one should set "portamento_max_time" to \fB4000\fR and "portamento_log_scale" to \fBtrue\fR to desired bus in the config file, bind CC5 to Portamento time and choose a monophonic mode in Surge
.IP "" 0
.SH "OSC CONTROLS"
.TP
\fB/play\fR
Start playback or restart if already playing
.TP
\fB/stop\fR
Stop playback
.TP
\fB/bpm\fR <float_or_int: bpm>
Set bpm
.TP
\fB/swing\fR <float_or_int: position>
Set swing strength (0: no swing, >0: swing, <0: anti\-swing)
.TP
\fB/swing/reference\fR <float_or_int: position>
Set swing reference (8: 8ths will swing, 16: 16th will swing, etc)
.TP
\fB/cursor\fR <float_or_int: position>
Set playhead position (affects all sequences)\. Position is >= 0 and expressed in quarter notes (0 = first beat)
.TP
\fB/screenset\fR <int: screen>
Change active screen set
.TP
\fB/panic\fR
Disable all sequences and cancel queued sequences
.TP
\fB/sequence\fR <string: mode> <int: column> <int: row>
Set sequence(s) state
.br
\fImode\fR: "solo", "on", "off", "toggle", "record", "record_on", "record_off", "sync", "clear", "copy", "cut", "paste", "delete"; only one sequence can be recording at a time; "record_off" mode doesn't require any argument
.br
\fIcolumn\fR: column number on screen set (zero indexed)
.br
\fIrow\fR: row number; if omitted, all rows are affected; multiple rows can be specified
.TP
\fB/sequence\fR <string: mode> <string: name>
Set sequence(s) state
.br
\fIname\fR: sequence name or osc pattern (can match multiple sequence names); multiple names can be specified
.TP
\fB/sequence/queue\fR <string: mode> <int: column> <int: row>
Same as /sequence but affected sequences will change state only on next cycle
.TP
\fB/sequence/trig\fR <string: mode> <int: column> <int: row>
Same as /sequence and (re)start playback
.TP
\fB/status\fR <string: address>
Send sequencer's status as json, without sequences informations
.br
\fIaddress\fR: \fIosc\.udp://ip:port\fR or \fIosc\.unix:///path/to/socket\fR ; if omitted the response will be sent to the sender
.TP
\fB/status/extended\fR <string: address>
Send sequencer's status as json, including sequences informations
.br

.SH "OSC STATUS"
.nf
{
    "screenset": \fIint\fR,
    "screensetName": "\fIstring\fR",
    "playing": \fIint\fR,
    "bpm": \fIint\fR,
    "tick": \fIint\fR,
    "sequences": [
        {
            "col": \fIint\fR,
            "row": \fIint\fR,
            "name": "\fIstring\fR",
            "time": "\fIstring\fR",
            "bars": \fIint\fR,
            "ticks": \fIint\fR,
            "queued": \fIint\fR,
            "playing": \fIint\fR,
            "timesPlayed": \fIint\fR,
            "recording": \fIint\fR
        },
        \|\.\|\.\|\.
    ]
}
.fi
.P
\fBSequencer status\fR
.IP "" 4
.nf
screenset: current screenset
screensetName: current screenset's name
playing: playback state
bpm: current bpm
tick: playback tick (192 ticks = 1 quarter note)
.fi
.IP "" 0
.P
\fBSequences statuses\fR (1 per active sequence in current screenset)
.IP "" 4
.nf
col: column position
row: row position
name: sequence name
time: sequence time signature (eg "4/4")
bars: number of bars in sequence
ticks: sequence length
queued: sequence's queued state
playing: sequence's playing state
timesPlayed: number of times the sequence played since last enabled
recording: sequence's recording state
.fi
.IP "" 0
.SH "AUTHORS"
seq192 is written by Jean\-Emmanuel Doucet and based on
.TP
seq24, written by
Rob C\. Buse, Ivan Hernandez, Guido Scholz, Dana Olson, Jaakko Sipari, Peter Leigh, Anthony Green, Daniel Ellis, Sebastien Alaiwan, Kevin Meinert, Andrea delle Canne
.TP
seq32, written by
Stazed
.SH "COPYRIGHT"
Copyright \(co 2021\-2025 Jean\-Emmanuel Doucet \fIjean\-emmanuel@ammd\.net\fR
.P
This program is free software: you can redistribute it and/or modify it under the terms of the GNU General Public License as published by the Free Software Foundation, either version 3 of the License, or (at your option) any later version\.
.P
This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE\. See the GNU General Public License for more details\.
.P
You should have received a copy of the GNU General Public License along with this program\. If not, see \fIhttps://www\.gnu\.org/licenses/\fR\.
.SH "LINKS"
Sources: \fIhttps://github\.com/jean\-emmanuel/seq192\fR


CHANGELOG.md
# Changelog

## 1.7.0

- main window
    - append "*" to window title when there are unsaved changes

- edit window
    - add `ctrl+s` shortcut to save the session from this window
    - fix shrinking / moving / pasting accross the sequence's end line
    - pianoroll visual tweaks

- sequence
    - add (FL-ish) slide notes

- config file
    - per bus "portamento_max_time" and "portamento_log_scale" settings for slide notes

## 1.6.1

- bug fixes
    - pitch wheel chasing not working when value LSB is 0 (always the case when set from pianoroll ui)

## 1.6.0

- edit window
    - save and restore grid unit / note length per sequence

- misc
    - now requires liblo 0.32

## 1.5.1

- nsm
  - fix save command in headless mode
  - don't announce optional gui capability in headless mode

## 1.5.0

- edit window
    - add measures and time signature to undo history
    - fix some unconsistent undo/redo behaviors

- misc
    - add osc controllable swing option

## 1.4.0

- main window
    - add undo/redo history for sequence grid and screenset names

- edit window
    - add a toggle button and a keyboard shortcut for switching between draw and select modes
    - fix some regressions with editing actions
    - prevent adding overlapping events when drawing multiple events in a single drag of mouse
    - add shift + click range selection
    - include sequence name changes in undo/redo history

- nsm
    - return correct error code when failing to save

## 1.3.0

- main window
    - add keyboard shortcuts for sequence menu actions
    - add `rename` action to sequence menu
    - add keyboard navigation in the sequence grid and a visual hint on the focused sequence
- edit window
    - when no event is selected, hitting Del or Backspace removes the event under the pointer
    - prevent crash when opening event menu with a config file loaded
    - fix display of user-defined instrument names in bus menu
- osc
    - add `/cursor` method
- sequence
    - synchronize the sequence when /sequence/queue is received, even if wanted playing state is the same than the current one
- misc
    - add `ctrl+space` shortcut to stop playback

## 1.2.0

- main window
    - re-enable mousewheel support on bpm and screenset entries and buttons
- edit window
    - display user-defined bus/channel color under the output menu button
- config file
    - allow setting colors, controls and note names per bus, not only per channel
- misc
    - various style tweaks

## 1.1.0

- main window
    - support user instrument colors (set in config file)
    - display user bus name and instrument name in sequence grid
    - add ctrl+click for queuing, shift+click for chosing sync reference
    - open edit window with middle-click on sequence

- perform
    - add the possibility to choose a sequence as sync reference for queued sequences (allows proper metric changes)

- sequence
    - add simple chasing option for control changes and pitch wheel events when sequence is disabled and when playback restarts : reset their value to 0 (or 8192 for pitch wheel) if the sequence has sent a different value while playing.

- edit window
    - fix upper bound of pitch wheel events: skip step 8064 to reach 8192 when displayed value is 127
    - change pitch wheel event slider's origin
    - chase option (under playback)
    - session not seen as modified when editing sequence
    - fix Aftertouch editing
    - center verticall scroll around present notes when the window opens
    - scroll vertically to follow transposition

- misc
    - new icons
    - seq192 can now be built without gtk (headless only)

## 1.0.0

- removed song editor
- interface rewritten with GTK3
- removed MIDI controls
- removed keyboard controls
- added OSC controls
- dropped support for systems other than GNU/Linux
- removed non-ALSA MIDI support
- starting transport while already playing resets position to 0
- added screenset export
- imported a bunch of features from seq32
- refactored config file system using JSON
- added resume playback mode to sequences
- added NSM support (with "optional-gui" and "dirty" capabilities)
- toggleable alt-control display in edit window
- 182 sequences per screenset


src/seq192.cpp
// This file is part of seq192
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.

#include <getopt.h>
#include <stdio.h>
#include <stdlib.h>
#include <signal.h>
#include <sys/stat.h>

#include "package.h"

#include "lib/nsm.h"

#include "core/midifile.h"
#include "core/configfile.h"
#include "core/cachefile.h"
#include "core/perform.h"

#ifdef USE_GTK
#include "gui/mainwindow.h"
#include <gtkmm.h>
#endif

/* struct for command parsing */
static struct
option long_options[] = {

    {"file",     required_argument, 0, 'f'},
    {"config", 1,0,'c'},
    {"help",     0, 0, 'h'},
    {"osc-port", 1,0,'p'},
    {"jack-transport",0, 0, 'j'},
    {"no-gui",0, 0, 'n'},
    {"version",0, 0, 'v'},
    {0, 0, 0, 0}

};

string config_filename = "";
string global_filename = "";
string last_used_dir = getenv("HOME");

#ifndef USE_GTK
bool global_no_gui = true;
#else
bool global_no_gui = false;
#endif

bool global_with_jack_transport = false;

bool global_is_running = true;

char* global_oscport;

string global_client_name = PACKAGE;

user_midi_bus_definition   global_user_midi_bus_definitions[c_maxBuses];
user_instrument_definition global_user_instrument_definitions[c_max_instruments];
user_keymap_definition     global_user_keymap_definitions[c_max_instruments];

#ifdef USE_GTK
color global_user_instrument_colors[c_max_instruments];
Glib::RefPtr<Gtk::Application> application;
#endif

// nsm
bool global_nsm_gui = false;
bool nsm_optional_gui_support = true;
nsm_client_t *nsm = 0;
bool nsm_wait = true;
string nsm_folder = "";
int
nsm_save_cb(char **,  void *userdata)
{
    perform *p = (perform *) userdata;
    if (p->file_save()) {
        return ERR_OK;
    } else {
        return ERR_GENERAL;
    }
}
void
nsm_hide_cb(void *userdata)
{
    #ifdef USE_GTK
    application->hold();
    #endif
    global_nsm_gui = false;
}
void
nsm_show_cb(void *userdata)
{
    global_nsm_gui = true;
}
int
nsm_open_cb(const char *name, const char *display_name, const char *client_id, char **out_msg, void *userdata)
{
    nsm_wait = false;
    nsm_folder = name;
    global_client_name = client_id;
    // NSM API 1.1.0: check if server supports optional-gui
    nsm_optional_gui_support = strstr(nsm_get_session_manager_features(nsm), "optional-gui");
    mkdir(nsm_folder.c_str(), 0777);
    // make sure nsm server doesn't override cached visibility state
    nsm_send_is_shown(nsm);
    return ERR_OK;
}


int
main (int argc, char *argv[])
{

    for (int i=0; i<c_maxBuses; i++)
    {
        for (int j=0; j<16; j++) {
            global_user_midi_bus_definitions[i].instrument[j] = -1;
            global_user_midi_bus_definitions[i].keymap[j] = -1;
        }
        global_user_midi_bus_definitions[i].portamento_max_time = 16383;
        global_user_midi_bus_definitions[i].portamento_log_scale = false;

    }

    for (int i=0; i<c_max_instruments; i++)
    {
        for (int j=0; j<128; j++)
            global_user_instrument_definitions[i].controllers_active[j] = false;
    }

    for (int i=0; i<c_max_instruments; i++)
    {
        for (int j=0; j<128; j++)
            global_user_keymap_definitions[i].keys_active[j] = false;
    }

    /* parse parameters */
    int c;
    while (1) {

        /* getopt_long stores the option index here. */
        int option_index = 0;

        c = getopt_long (argc, argv, "p:f:c:hjnv", long_options, &option_index);

        /* Detect the end of the options. */
        if (c == -1)
            break;

        switch (c) {

            case '?':
            case 'h':

                printf("\n");
                printf("%s - live MIDI sequencer (version %s)", PACKAGE, VERSION);

                #if !defined USE_JACK && !defined USE_GTK
                printf("(compiled without jack support and gtk support)");
                #elif !defined USE_JACK
                printf("(compiled without jack support)");
                #elif !defined USE_GTK
                printf("(compiled without gtk support)");
                #endif

                printf("\n\nUsage: %s [options]\n\n", PACKAGE);
                printf("Options:\n");
                printf("  -h, --help              show available options\n");
                printf("  -f, --file <filename>   load midi file on startup\n");
                printf("  -c, --config <filename> load config file on startup\n");
                printf("  -p, --osc-port <port>   osc input port (udp port number or unix socket path)\n");
                #ifdef USE_JACK
                printf("  -j, --jack-transport    sync to jack transport\n");
                #endif
                #ifdef USE_GTK
                printf("  -n, --no-gui            enable headless mode\n");
                #endif
                printf("  -v, --version           show version and exit\n");
                printf("\n");

                return 0;
                break;

            case 'j':
                global_with_jack_transport = true;
                break;

            case 'n':
                global_no_gui = true;
                break;

            case 'f':
                global_filename = string(optarg);
                break;

            case 'c':
                config_filename = string(optarg);
                break;

            case 'p':
                global_oscport = optarg;
                break;

            case 'v':
                printf("%s %s\n", PACKAGE, VERSION);
                return EXIT_SUCCESS;
                break;

            default:
                break;
        }

    }

    // nsm
    const char *nsm_url = getenv( "NSM_URL" );
    if (nsm_url) {
        nsm = nsm_new();
        nsm_set_open_callback(nsm, nsm_open_cb, 0);
        if (nsm_init(nsm, nsm_url) == 0) {
            if (!global_no_gui) nsm_send_announce(nsm, PACKAGE, ":optional-gui:dirty:", argv[0]);
            else nsm_send_announce(nsm, PACKAGE, ":dirty:", argv[0]);
        }
        int timeout = 0;
        while (nsm_wait) {
            nsm_check_wait(nsm, 500);
            timeout += 1;
            if (timeout > 200) exit(1);
        }
    }

    /* the main performance object */
    perform * p = new perform();

    // read config file
    string config_path = getenv("XDG_CONFIG_HOME") == NULL ? string(getenv("HOME")) + "/.config" : getenv("XDG_CONFIG_HOME");
    config_path += string("/") + PACKAGE;
    mkdir(config_path.c_str(), 0777);
    if (nsm) config_path = nsm_folder;
    string file_path = config_filename == "" ? (config_path + "/config.json") : config_filename;
    std::ifstream infile(file_path);
    if (!infile.good()) {
        std::fstream fs;
        fs.open(file_path, std::ios::out);
        fs << "{}" << endl;
        fs.close();
    }
    ConfigFile config(file_path);
    config.parse();

    // read/touch cache file
    string cache_path = getenv("XDG_CACHE_HOME") == NULL ? string(getenv("HOME")) + "/.cache" : getenv("XDG_CACHE_HOME");
    cache_path += string("/") + PACKAGE;
    mkdir(cache_path.c_str(), 0777);
    if (nsm) cache_path = nsm_folder;
    file_path = cache_path + "/cache.json";
    infile = std::ifstream(file_path);
    if (!infile.good()) {
        std::fstream fs;
        fs.open(file_path, std::ios::out);
        fs << "{}" << endl;
        fs.close();
    }
    CacheFile cache(file_path);
    cache.parse();

    p->init();

    p->launch_input_thread();
    p->launch_output_thread();

    #ifdef USE_JACK
    p->init_jack();
    #endif

    if (nsm) {
        global_filename = nsm_folder + "/session.midi";
        // write session file if it doesn't exist
        std::ifstream infile(global_filename);
        if (!infile.good()) {
            midifile f(global_filename);
            f.write(p, -1, -1);
        }
        nsm_set_save_callback(nsm, nsm_save_cb, (void*) p);
    }

    if (global_filename != "") {
        midifile *f = new midifile(global_filename);
        f->parse(p, 0);
        delete f;
    }

    if (global_oscport != 0 && strstr(global_oscport, "/")) {
        // liblo needs a gracefull ctrl+c handler to release unix socket
        signal(SIGINT, [](int param){global_is_running = false;});
    }

    int status = 0;
    if (global_no_gui) {
        while (global_is_running) {
            if (nsm) nsm_check_nowait(nsm);
            usleep(10000);
        }
    } else {
        #ifdef USE_GTK
        application = Gtk::Application::create();
        MainWindow window(p, application);

        if (nsm) {
            // setup optional-gui
            if (nsm_optional_gui_support) {
                nsm_set_show_callback(nsm, nsm_show_cb, 0);
                nsm_set_hide_callback(nsm, nsm_hide_cb, 0);
                if (!global_nsm_gui) {
                  window.hide();
                  nsm_send_is_hidden(nsm);
                }
                else nsm_send_is_shown(nsm);
            } else {
                global_nsm_gui = true;
            }
            // enable nsm in window
            window.nsm_set_client(nsm, nsm_optional_gui_support);
            // bind quit signal
            signal(SIGTERM, [](int param){
                global_is_running = false;
                application->quit();
            });
        }

        // bind ctrl+c signal
        signal(SIGINT, [](int param){
            global_is_running = false;
            application->quit();
        });


        status = application->run(window);
        #endif
    }

    // write cache file
    cache.write();

    delete p;

    if (nsm) {
        nsm_free(nsm);
        nsm = NULL;
    }

    return status;
}
